services/src/main/java/io/scalecube/services/ServiceRegistry.java

```diff
@@ -0,0 +1,181 @@
+        .forEach(entry -> {
+          ServiceReference serviceRef = new ServiceReference(
+              member.id(),
+              entry.getKey(),
+              member.address());
+          
+          LOGGER.debug("Member: {} is {} : {}", member, type, serviceRef);
+          if (type.equals(DiscoveryType.ADDED) || type.equals(DiscoveryType.DISCOVERED)) {
+            serviceInstances.putIfAbsent(serviceRef, new RemoteServiceInstance(cluster, serviceRef));
+          } else if (type.equals(DiscoveryType.REMOVED)) {
+            serviceInstances.remove(serviceRef);
+          }
+        });
+  }
+
+  /**
+   * register a service instance at the cluster.
+   */
+  public void registerService(Object serviceObject) {
+    checkArgument(serviceObject != null, "Service object can't be null.");
+    Collection<Class<?>> serviceInterfaces = serviceProcessor.extractServiceInterfaces(serviceObject);
+
+    String memberId = cluster.member().id();
+
+    serviceInterfaces.forEach(serviceInterface -> {
+      // Process service interface
+      Map<String, ServiceDefinition> serviceDefinitions =
+          serviceProcessor.introspectServiceInterface(serviceInterface);
+
+      serviceDefinitions.values().forEach(definition -> {
+        ServiceReference serviceRef = new ServiceReference(memberId, definition.qualifier(), cluster.address());
+        ServiceInstance serviceInstance =
+            new LocalServiceInstance(serviceObject, memberId, definition.qualifier(), definition.method());
+        serviceInstances.putIfAbsent(serviceRef, serviceInstance);
+
+      });
+    });
+  }
+
+  @Override
+  public void unregisterService(Object serviceObject) {
+    checkArgument(serviceObject != null, "Service object can't be null.");
+    Collection<Class<?>> serviceInterfaces = serviceProcessor.extractServiceInterfaces(serviceObject);
+
+    serviceInterfaces.forEach(serviceInterface -> {
+      // Process service interface
+      Map<String, ServiceDefinition> serviceDefinitions =
+          serviceProcessor.introspectServiceInterface(serviceInterface);
+
+      serviceDefinitions.values().forEach(serviceDefinition -> {
+        ServiceReference serviceReference = toLocalServiceReference(serviceDefinition);
+        serviceInstances.remove(serviceReference);
+      });
+    });
+  }
+
+  @Override
+  public List<ServiceInstance> serviceLookup(final String serviceName) {
+    checkArgument(serviceName != null, "Service name can't be null");
+    return serviceInstances.entrySet().stream()
+        .filter(entry -> isValid(entry.getKey(), serviceName))
+        .map(Map.Entry::getValue)
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public Optional<ServiceInstance> getLocalInstance(String serviceName) {
+    return Optional.ofNullable(serviceInstances.values().stream()
+        .filter(ServiceInstance::isLocal)
+        .filter(serviceInstance -> serviceInstance.serviceName().equals(serviceName))
+        .findFirst()
+        .orElse(null));
```

Comment:
findFirst already return optional, you don't need to create it from nullable and put null in orElse